test/pinverse.cpp

```diff
@@ -0,0 +1,327 @@
+    array u;
+    array vT;
+    array sVec;
+    svd(u, sVec, vT, in);
+    dim_t sSize = sVec.elements();
+
+    sVec(2) = (double) 1e-16;
+    array s = diag(sVec, 0, false);
+    array zeros = constant(0,
+                           dim0 > sSize ? dim0 - sSize : sSize,
+                           dim1 > sSize ? dim1 - sSize : sSize,
+                           f64);
+    s = join(dim0 > dim1 ? 0 : 1, s, zeros);
+
+    // Make new input array that has a small non-zero value in its SVD sigma
+    in = matmul(u, s, vT);
+    array inpinv = pinverse(in, 1e-15);
+    array out = matmul(in, inpinv, in);
+
+    ASSERT_ARRAYS_NEAR(in, out, eps<double>());
+}
+
+TEST(Pinverse, Batching) {
+    array in = readTestInput<float>(string(TEST_DIR"/pinverse/pinverse10x8x2.test"));
+    array inpinv0 = pinverse(in(span, span, 0));
+    array inpinv1 = pinverse(in(span, span, 1));
+
+    array out = pinverse(in);
+    array out0 = out(span, span, 0);
+    array out1 = out(span, span, 1);
+
+    ASSERT_ARRAYS_NEAR(inpinv0, out0, eps<float>());
+    ASSERT_ARRAYS_NEAR(inpinv1, out1, eps<float>());
+}
+
+TEST(Pinverse, CustomTol) {
+    array in = readTestInput<float>(string(TEST_DIR"/pinverse/pinverse10x8.test"));
+    array inpinv = pinverse(in, 1e-12);
+    array out = matmul(in, inpinv, in);
+    ASSERT_ARRAYS_NEAR(in, out, eps<float>());
+}
+
+TEST(Pinverse, C) {
+    array in = readTestInput<float>(string(TEST_DIR"/pinverse/pinverse10x8.test"));
+    af_array inpinv = 0, out = 0;
+    af_pinverse(&inpinv, in.get(), 1e-6, AF_MAT_NONE);
+    af_matmul(&out, in.get(), inpinv, AF_MAT_NONE, AF_MAT_NONE);
+    af_matmul(&out, out, in.get(), AF_MAT_NONE, AF_MAT_NONE);
+
+    ASSERT_ARRAYS_NEAR(in.get(), out, eps<float>());
+
+    af_release_array(out);
+    af_release_array(inpinv);
+}
+
+TEST(Pinverse, C_CustomTol) {
+    array in = readTestInput<float>(string(TEST_DIR"/pinverse/pinverse10x8.test"));
+    af_array inpinv = 0, out = 0;
+    af_pinverse(&inpinv, in.get(), 1e-12, AF_MAT_NONE);
+    af_matmul(&out, in.get(), inpinv, AF_MAT_NONE, AF_MAT_NONE);
+    af_matmul(&out, out, in.get(), AF_MAT_NONE, AF_MAT_NONE);
+
+    ASSERT_ARRAYS_NEAR(in.get(), out, eps<float>());
+
+    af_release_array(out);
+    af_release_array(inpinv);
+}
+
+TEST(Pinverse, NegativeTol) {
+    array in = readTestInput<float>(string(TEST_DIR"/pinverse/pinverse10x8.test"));
+    array out;
+    ASSERT_THROW(out = pinverse(in, -1.f), exception);
+}
+
+TEST(Pinverse, InvalidType) {
+    array in = constant(0, 10, 8, u8);
+    array out;
+    ASSERT_THROW(out = pinverse(in, -1.f), exception);
+}
+
+TEST(Pinverse, TooManyDims) {
+    array in = constant(0.f, 10, 8, 2, f32);
+    array out;
+    ASSERT_THROW(out = pinverse(in, -1.f), exception);
+}
+
+TEST(Pinverse, InvalidMatProp) {
+    array in = constant(0.f, 10, 8, f32);
+    array out;
+    ASSERT_THROW(out = pinverse(in, -1.f, AF_MAT_SYM), exception);
+}
+
+TEST(Pinverse, DocSnippet) {
+    float hA[] = {0, 1, 2, 3, 4, 5};
```

Comment:
This array should be part of the example. Otherwise, the user doesn't have any context for hA.

Reply:
Oh whoops yup let me include it